### ISSUE TITLE
relinking library directory should automatically rescan library

### ISSUE THREAD

**Issue by Be-ing** [mixxxdj/mixxx#12858]
### Bug Description

Relinking a library directory in the Library preferences requires manually rescanning the library for Mixxx to find the files. This should be done automatically when saving the changed preferences.

### Version

2.4.0

### OS

Fedora 39

**Comment by manuelbertelli** [comment]
How do I assign this issue to me?

**Comment by Be-ing** [comment]
There's no need to assign the issue, you can just go ahead and open a pull request when you have a fix implemented.

**Comment by manuelbertelli** [comment]
Relinking = relocate dir?

**Comment by Be-ing** [comment]
Yes, using the Relink button in the Library preferences.

**Comment by ronso0** [comment]
Why is a rescan _required_?

IIUC, if the (inner) directory structure didn't change (dirs or tracks renamed / deleted) all tracks should be updated correctly.
or are you referring to the relink after migrating the Mixxx db to another OS / computer / ...? In that case encoders (availability / version) might have changed and rescan might be needed.

**Comment by ronso0** [comment]
Note: relinking was actually broken before #12462 (dirs were update but not the tracks).

**Comment by Be-ing** [comment]
> are you referring to the relink after migrating the Mixxx db to another OS / computer / ...?

I noticed this issue when testing migration into the Flatpak sandbox https://mixxx.org/news/2024-02-21-official-flatpak-now-available-on-flathub/

**Comment by ronso0** [comment]
I saw that, but does that mean we need a full rescan in order to get access to each track?

**Comment by ronso0** [comment]
Okay, I tried this myself:
* installed org.mixxx.Mixxx
* (~/.mixxx exists but is empty)
* copied _mixxx.cfg_ and _mixxxdb.sqlite_ to _~/.var/app/org.mixxx.Mixxx/.mixxx_
* start Mixxx with `flatpak run org.mixxx.Mixxx`
* asks me for a music directory (mixxxdb.sqlite and config are ignored?), I pick one
* close Mixxx
* now I suddenly have _~/.mixxx_
* okay, so I copy _mixxx.cfg_ and _mixxxdb.sqlite_ there
* start Mixxx, old music dirs are shown in Pref -> Library
* relink one that is marked missing: no permission, no advice how to grant it
* close Mixxx, research how to grant file permissions, do so with
`sudo flatpak override org.mixxx.Mixxx --filesystem=~/someMusicDir`
(I also used `Flatseal` for another directory)
* start Mixxx
* all tracks in _~/someMusicDir_ are accessible

### Conclusion:
The rescan is not required in general. Relink is sufficient.
I did this test on Ubuntu 20.04.05. If this works differently on other OS, okay, but I don't agree we should demand the rescan until we figured when/why it is necessary.

**Comment by Be-ing** [comment]
`flatpak override org.mixxx.Mixxx --filesystem=~/someMusicDir`

That will change the result. That shouldn't be required.

**Comment by ronso0** [comment]
The only useful dir permissions Mixxx has after installation are for `/media` and `xdg-music`. My main music directory is on another drive mounted in `/mnt`, and one in `/home`. In the Relink file picker there is no way to see other directories, let alone grant access. Neither https://flathub.org/apps/org.mixxx.Mixxx nor https://github.com/flathub/org.mixxx.Mixxx give any hints how to deal with this.

The [Flatpak sandbox docs](https://docs.flatpak.org/en/latest/sandbox-permissions-reference.html#filesystem-permissions) advice to use `flatpak override [app-id] --filesystem=~/file/path`.
As I wrote above, I also used **`Flatseal`** to grant access to another directory: no relink required, no issues.
(FWIW Flatseal did not list the dir I added previously with `flatpak override org.mixxx.Mixxx --filesystem=`)
When I test with a music dir in `/media`: no relink required, no issues.

This is with 
`flatpak-1.6.5-0ubuntu.4`
`xdg-desktop-portal-1.6.1-1ubuntu2`

**Comment by Be-ing** [comment]
> In the Relink file picker there is no way to see other directories

Really? The file picker dialog goes through the XDG File Chooser portal, so it should let you choose any directory that your user has permission to.

**Comment by Be-ing** [comment]
> The only useful dir permissions Mixxx has after installation are for /media and xdg-music.

Right, relinking is only needed if the user already had directories outside of these paths in their Mixxx database.

**Comment by ronso0** [comment]
> > In the Relink file picker there is no way to see other directories
> 
> Really? The file picker dialog goes through the XDG File Chooser portal, so it should let you choose any directory that your user has permission to, including outside the Flatpak sandbox.

What shall I say, xdg-desktop-portal  is **1.6.1-1ubuntu2** which is quite old I guess.
Anyhow, please simulate the migration yourself again and verify a rescan is indeed still required.

**Comment by Toneplay** [comment]
Would rescanning after a Relink ever be a terrible thing? Wouldn't it be fast if it isn't necessary (that is, if the records for the relinked folder are already in the library)?

Maybe after Relinking, rather than forcing a rescan, we should make rescanning optional and prompted, just as we do when the user Adds a new directory.

It seems wrong not to warn the user that the directory they just relinked might not be available until rescan.

**Comment by ronso0** [comment]
> It seems wrong not to warn the user that the directory they just relinked might not be available until rescan.

The dir is available immediately, relinking just replaces the file paths in the database which is blazing fast.
If that doesn't work please file a bug report.
(Note: this was not the case when relinking was broken, so I suggest you try again with Mixxx 2.5, or at least 2.4.1 where I fixed it)

> Would rescanning after a Relink ever be a terrible thing?

Define 'terrible' ; )
Thing is, currently we can only rescan ALL directories. Directories can be
* huge
* on various partitions, incl. slow HDD

Rescan blocks the GUI (to avoid database inconsistencies caused by concurrent access (scanner vs. user modifications like metadata edit, play count etc.) and it can take long.
So, considering it's not _necessary_ for relinking I vote for not doing it automatically.

Refactoring the scanner to allow scanning specific directories is an option though, as well as offering a specific rescan after relinking.
Related:
#13989 
#14142
#8532

**Comment by Toneplay** [comment]
> The dir is available immediately,

It is not, if rescanning is enabled at startup and you just started Mixxx. The rescan removes the entries, so it's just like a "Remove".

**Comment by ronso0** [comment]
> It is not, if rescanning is enabled at startup and you just started Mixxx. The rescan removes the entries, so it's just like a "Remove" in terms of library contents.

... and auto-rescan bites us again :disappointed: 
In that case Move tracks -> start Mixxx -> auto-rescan another rescan is required after relinking.
For the reasons mentioned above rescaning only affected directoies would be preferrable.

**Comment by ronso0** [comment]
I've adjusted the bug title according to reflect our findings.
